### ISSUE TITLE
problem plotting after clean installation

### ISSUE THREAD

**Issue by AEljarrat** [hyperspy/hyperspy#1950]
Dear all, 
its been a while since i've been able to make an installation of hyperspy with working plot capabilities from scratch. Following the steps in the documentation, I install hyperspy in a clean conda environment (python 3, cython) using pip and the source code from git. Typically, I get a PyQT error when trying to use plots. Finally, I've found this error is solved if I additionally install pyside2; `conda install pyside2 -c conda-forge`.

I don't know if there is a reason why this package is not automatically installed as a dependency of hyperspy. Anyhow, I thought it would maybe be helpful to add this info to the documentation. I am not sure about this, but I guess it does not hurt to report it, so we can discuss further.

Cheers!
Alberto.

**Comment by ericpre** [comment]
What is the error you are getting exactly?
If you install hyperspy using pip on a bare python environment, you will have the default tk python gui toolkit as backend for matplotlib and pyqt will not be installed. I think this is the intended behaviour because hyperspy should not force the installation of a particular gui toolkit.
```python
%matplotlib
import hyperspy.api as hs
```
If you want to use pyqt, you will need to install it by using `pip install pyqt5` for example.

```python
%matplotlib qt
import hyperspy.api as hs
```
If I remember correctly, pyside2 is not supported by pyface which can lead to some error/warning using guis.

There are a few things outdated in the installation section of the user guide and indeed there is room for improvement regarding installation from source.

**Comment by AEljarrat** [comment]
Thanks for the explanation. 

It seems I had not researched the issue properly, and this has to do more with me not understanding how to use the matplotlib backends.

In the clean installation I can use the tk backend,
```python
%matplotlib tk
import hyperspy.api as hs
```
This works just fine. If I don't invoke tk explicitly I get an error, similar to the one I get when trying to invoke qt (see Error 1 and 2 below). Indeed, for version matplotlib 2.2 installing pyqt5 does the trick. I think pyqt4 should be used for some older versions of matplotlib. Additionally, PySide works too and I haven't observed any weird warning/error related (disclaimer, I'm not using GUIs so far). 

Maybe the installation section of the user guide could be more precise and explain that tk functionality is available from scratch (I understand it is the default backend) and/or the additional step necessary to obtain plotting capabilities with qt. 

I want to try to use the GUIs before proposing any addition to the user guide, so that I am sure what I'm doing.

### Error 1
```python
%matplotlib qt
import matplotlib.pyplot as plt
import numpy as np
import hyperspy.api as hs

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/qt_compat.py in <module>()
    200     try:
--> 201         from PySide import QtCore, QtGui, __version__, __version_info__
    202     except ImportError:

ModuleNotFoundError: No module named 'PySide'

During handling of the above exception, another exception occurred:

ImportError                               Traceback (most recent call last)
<ipython-input-1-344815c28dbc> in <module>()
----> 1 get_ipython().run_line_magic('matplotlib', 'qt')
      2 #import matplotlib
      3 #matplotlib.use('QT5Agg')
      4 import matplotlib.pyplot as plt
      5 import numpy as np

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/interactiveshell.py in run_line_magic(self, magic_name, line, _stack_depth)
   2129                 kwargs['local_ns'] = sys._getframe(stack_depth).f_locals
   2130             with self.builtin_trap:
-> 2131                 result = fn(*args,**kwargs)
   2132             return result
   2133

<decorator-gen-107> in matplotlib(self, line)

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/magic.py in <lambda>(f, *a, **k)
    185     # but it's overkill for just that one bit of state.
    186     def magic_deco(arg):
--> 187         call = lambda f, *a, **k: f(*a, **k)
    188
    189         if callable(arg):

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/magics/pylab.py in matplotlib(self, line)
     97             print("Available matplotlib backends: %s" % backends_list)
     98         else:
---> 99             gui, backend = self.shell.enable_matplotlib(args.gui)
    100             self._show_matplotlib_backend(args.gui, backend)
    101

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/interactiveshell.py in enable_matplotlib(self, gui)
   3049                 gui, backend = pt.find_gui_and_backend(self.pylab_gui_select)
   3050
-> 3051         pt.activate_matplotlib(backend)
   3052         pt.configure_inline_support(self, backend)
   3053

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/pylabtools.py in activate_matplotlib(backend)
    308     matplotlib.rcParams['backend'] = backend
    309
--> 310     import matplotlib.pyplot
    311     matplotlib.pyplot.switch_backend(backend)
    312

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/pyplot.py in <module>()
    113 ## Global ##
    114
--> 115 _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
    116
    117 _IP_REGISTERED = None

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/__init__.py in pylab_setup(name)
     60     # imports. 0 means only perform absolute imports.
     61     backend_mod = __import__(backend_name, globals(), locals(),
---> 62                              [backend_name], 0)
     63
     64     # Things we pull in from all backends

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/backend_qt5agg.py in <module>()
     13
     14 from .backend_agg import FigureCanvasAgg
---> 15 from .backend_qt5 import (
     16     QtCore, QtGui, QtWidgets, _BackendQT5, FigureCanvasQT, FigureManagerQT,
     17     NavigationToolbar2QT, backend_version)

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/backend_qt5.py in <module>()
     17     _Backend, FigureCanvasBase, FigureManagerBase, NavigationToolbar2,
     18     TimerBase, cursors, ToolContainerBase, StatusbarBase)
---> 19 import matplotlib.backends.qt_editor.figureoptions as figureoptions
     20 from matplotlib.backends.qt_editor.formsubplottool import UiSubplotTool
     21 from matplotlib.figure import Figure

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/qt_editor/figureoptions.py in <module>()
     18 import matplotlib
     19 from matplotlib import cm, colors as mcolors, markers, image as mimage
---> 20 import matplotlib.backends.qt_editor.formlayout as formlayout
     21 from matplotlib.backends.qt_compat import QtGui
     22

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/qt_editor/formlayout.py in <module>()
     52
     53 from matplotlib import colors as mcolors
---> 54 from matplotlib.backends.qt_compat import QtGui, QtWidgets, QtCore
     55
     56

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/qt_compat.py in <module>()
    202     except ImportError:
    203         raise ImportError(
--> 204             "Matplotlib qt-based backends require an external PyQt4, PyQt5,\n"
    205             "PySide or PySide2 package to be installed, but it was not found.")
    206

ImportError: Matplotlib qt-based backends require an external PyQt4, PyQt5,
PySide or PySide2 package to be installed, but it was not found.
```
### Error 2
```python
%matplotlib
import matplotlib.pyplot as plt
import numpy as np
import hyperspy.api as hs

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-2-197202344344> in <module>()
----> 1 get_ipython().run_line_magic('matplotlib', '')
      2 #import matplotlib
      3 #matplotlib.use('QT5Agg')
      4 import matplotlib.pyplot as plt
      5 import numpy as np

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/interactiveshell.py in run_line_magic(self, magic_name, line, _stack_depth)
   2129                 kwargs['local_ns'] = sys._getframe(stack_depth).f_locals
   2130             with self.builtin_trap:
-> 2131                 result = fn(*args,**kwargs)
   2132             return result
   2133

<decorator-gen-107> in matplotlib(self, line)

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/magic.py in <lambda>(f, *a, **k)
    185     # but it's overkill for just that one bit of state.
    186     def magic_deco(arg):
--> 187         call = lambda f, *a, **k: f(*a, **k)
    188
    189         if callable(arg):

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/magics/pylab.py in matplotlib(self, line)
     97             print("Available matplotlib backends: %s" % backends_list)
     98         else:
---> 99             gui, backend = self.shell.enable_matplotlib(args.gui)
    100             self._show_matplotlib_backend(args.gui, backend)
    101

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/interactiveshell.py in enable_matplotlib(self, gui)
   3049                 gui, backend = pt.find_gui_and_backend(self.pylab_gui_select)
   3050
-> 3051         pt.activate_matplotlib(backend)
   3052         pt.configure_inline_support(self, backend)
   3053

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/IPython/core/pylabtools.py in activate_matplotlib(backend)
    308     matplotlib.rcParams['backend'] = backend
    309
--> 310     import matplotlib.pyplot
    311     matplotlib.pyplot.switch_backend(backend)
    312

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/pyplot.py in <module>()
    113 ## Global ##
    114
--> 115 _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
    116
    117 _IP_REGISTERED = None

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/__init__.py in pylab_setup(name)
     60     # imports. 0 means only perform absolute imports.
     61     backend_mod = __import__(backend_name, globals(), locals(),
---> 62                              [backend_name], 0)
     63
     64     # Things we pull in from all backends

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/backend_qt5agg.py in <module>()
     13
     14 from .backend_agg import FigureCanvasAgg
---> 15 from .backend_qt5 import (
     16     QtCore, QtGui, QtWidgets, _BackendQT5, FigureCanvasQT, FigureManagerQT,
     17     NavigationToolbar2QT, backend_version)

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/backend_qt5.py in <module>()
     17     _Backend, FigureCanvasBase, FigureManagerBase, NavigationToolbar2,
     18     TimerBase, cursors, ToolContainerBase, StatusbarBase)
---> 19 import matplotlib.backends.qt_editor.figureoptions as figureoptions
     20 from matplotlib.backends.qt_editor.formsubplottool import UiSubplotTool
     21 from matplotlib.figure import Figure

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/qt_editor/figureoptions.py in <module>()
     18 import matplotlib
     19 from matplotlib import cm, colors as mcolors, markers, image as mimage
---> 20 import matplotlib.backends.qt_editor.formlayout as formlayout
     21 from matplotlib.backends.qt_compat import QtGui
     22

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/qt_editor/formlayout.py in <module>()
     52
     53 from matplotlib import colors as mcolors
---> 54 from matplotlib.backends.qt_compat import QtGui, QtWidgets, QtCore
     55
     56

~/anaconda3/envs/hyperspy_mock/lib/python3.6/site-packages/matplotlib/backends/qt_compat.py in <module>()
    138     if QT_API == QT_API_PYQT5:
    139         try:
--> 140             from PyQt5 import QtCore, QtGui, QtWidgets
    141             _getSaveFileName = QtWidgets.QFileDialog.getSaveFileName
    142         except ImportError:

ModuleNotFoundError: No module named 'PyQt5'
```

**Comment by ericpre** [comment]
Recently, I checked installing hyperspy with pip in a bar python environment and there are still (at least) two of the libraries (statsmodels and traits), which requires compiling C code during the installation process.
The matplotlib backend setting issue that you are mentioning here may depends on the matplotlib configuration file which in turn may depend on the python distribution and it may also depends on other/previous python installation...
In short, I don't think that we can recommend a _robust_ hyperspy installation from source which will work all the time. As it was suggested by others previously, one easy way to get started with a development version is to use miniconda, install hyperspy from conda-forge, uninstall it and then install hyperspy from source. Even if it sounds a bit silly, maybe this is one way that we can recommend in the user guide?

**Comment by AEljarrat** [comment]
Installing and uninstalling from conda-forge, and then installing development works. Maybe we could recommend doing 
```python
conda install hyperspy --only-deps -c conda-forge
pip install -e ./
```
So that only the dependencies are installed, but we save the hassle of uninstalling. This works for me on windows, I want to try this also on linux, but I won't be able until next week. What do you think?

**Comment by ericpre** [comment]
Yes it sounds good!
